Pave the way for converting nRF devicetrees to use phandles for pinmuxing

mbolivar-nordic opened the pull request:
RFC for now because of these questions:

1. what "flags" value should we use for I2C `sda-gpios` and `scl-gpios`? The driver ignores them, so I guess maybe documenting that and saying "just put 0" is right, even if it's a bit inconsistent with Linux's use of those properties for bit-banged I2C.
2. once we settle that, I need to convert all in-tree devicetrees to stop using the deprecated properties

Nordic SoC devices do their pinmuxing in devicetree using properties that look like `foo-pin = <PSEL value>;`, where a "PSEL value" is an integer that represents a GPIO (like P0.1, P1.2) to a bitmask with the lower 5 bits representing the pin (0-31), and the 6th bit representing the port (0 or 1).

This requires users to manually encode and decode pins to ints and vice versa.

This has two problems:

- it confuses people who aren't familiar with the underlying register maps (which usually have registers that take the PSEL values directly)
- it "hides" device dependencies on GPIO controllers, which may break automated tooling related to the devicetree, such as dependency tracking done for things like power management (#29644) or generating pin maps without special cases (https://github.com/zephyrproject-rtos/zephyr/pull/24132)

The same solution applies to both problems: use standard devicetree GPIO phandle and specifier, so you could say:

```
foo-gpios = <&gpio0 1 FLAGS>;   // P0.1
```

or

```
foo-gpios = <&gpio1 2 FLAGS>; // P1.2
```

This pull request makes that possible with backwards compatibility and sensible error messages from drivers that convert to the new style:

- both `foo-pin` and `foo-gpios` are supported, but `foo-pin` can be deprecated
- if both `foo-pin` and `foo-gpios` are given in a devicetree, the user gets a readable error message identifying the node and properties that need to be fixed
- the same is true if neither `foo-pin` nor `foo-gpios` is given

This is shown using the I2C drivers. I tested this using a sensor driver with the new `sda-gpios` and `scl-gpios` properties on nRF52840-DK.

mbolivar-nordic committed:
doc: dts: add link from bindings guide to bindings index

We recently added a generated devicetree bindings index. Link to it
from the bindings page in the Guides section of the documentation so
it's easier for beginners to find.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: dts: gen_defines: don't include descriptions

Before we had a bindings index in the documentation, the generated
header file was (somewhat unfortunately) often our best reference for
what a particular binding or property within a binding ends up doing,
so it made good sense to put the description in the generated file.

Now that we have HTML documentation that's a bit more digestible than
the generated file, though, we can just point users at that. Do that
and remove the inline description from the generated file.

This makes it possible to put C-style multiline comments in the
descriptions themselves, which will be done in subsequent patches.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: bindings: improve documentation for Nordic IPs

Add some details to help new users learn how to manage the nodes
for compatibles corresponding to IP blocks on Nordic SoCs.

The focus is on IP blocks where the pinmux is configured in
devicetree. These typically have individual properties for each pin,
like a SCK pin on a SPI peripheral, which correspond directly to a
PSEL register value. These violate the usual devicetree convention of
using a 'foo-gpios <&gpioX ...>' phandle property and work in ways
that only make sense if you understand the underlying register map, so
it's worth explaining them a bit more carefully.

Skip these two bindings, which have no drivers: nordic,nrf-i2s.yaml,
nordic,nrf-pdm.yaml.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
soc: nrf: clean up soc.h headers

Make a common header for common includes.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
soc: nrf: add DT GPIO helper macros

Add some helper macros that will be convenient to use from device
drivers for accessing and error checking pin mux information in the
devicetree:

- NRF_DT_GPIOS_TO_PSEL(): converts a devicetree GPIO phandle
  <&gpioX pin flags> to the "PSEL style" registers found in
  nRF SoC family IPs. The conversion is done at
  build time and relies on devicetree properties.

- NRF_DT_CHECK_GPIO_CTLR_IS_SOC(): make sure a GPIO controller
  in a phandle-array property is one of the SoC GPIO controllers
  by checking its compatible is nordic,nrf-gpio.

See comments in the patch for more details.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: bindings: spi-controller.yaml documentation

Add more explanations etc.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
devicetree: add DT_NODE_PATH()

This macro returns a node's full path, given its node identifier.

The entire path to a node is useful information for the user which can
be added to build-time error messages.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
Once we get this out of RFC, I'll tack on more patches converting the other drivers and updating in-tree DTS files. My guess is this won't make 2.5 but I would like to get it into 2.6.

pabigot commented on scripts/dts/gen_defines.py:
```diff
@@ -195,11 +195,19 @@ def write_node_comment(node):
+        # inconvenient when we want to do things like quote snippets
+        # of .dtsi files within the snippets themselves, or otherwise
+        # include the characters "*/" in order.
+        s += ("\n(Descriptions have moved to the Devicetree Bindings Index\n"
```
FWIW I had tried to include comments in descriptions and it broke YAML parsing.  Might have been due to things that were within the comments, though, since it appears the ones you added work.

Also this will eliminate any in-header description for bindings that aren't in any generated documentation.

mbolivar-nordic replied:
> Also this will eliminate any in-header description for bindings that aren't in any generated documentation.

I don't think that's true. If a description from a binding makes it into the header, it's because the relevant node had a matching compatible. Which means that compatible's binding will be in the bindings index.

Unless I'm missing something, that can't happen.

pabigot replied:
I was thinking about out-of-tree bindings: application specific, or module-specific, or other situations where somebody might not have a comprehensive Devicetree Bindings Index (or wouldn't know to generate one).  But I haven't dug into what the tooling provides in that situation.

mbolivar-nordic replied:
> I was thinking about out-of-tree bindings: application specific, or module-specific, or other situations where somebody might not have a comprehensive Devicetree Bindings Index

Fair enough, but I think at that point they know the binding and this is a worthwhile tradeoff until we get better tooling for inspecting the DT.

pabigot replied:
I don't have a problem with the change; the generated comment still provides the full path to the binding, so people should have enough of a clue.

pabigot commented on soc/arm/nordic_nrf/soc_nrf_common.h:
```diff
@@ -13,6 +13,155 @@
+	 UTIL_AND(DT_NODE_EXISTS(DT_NODELABEL(gpio1)),			\
+		  NRF_DT_GPIO_CTLR_IS_GPIOX(node_id, prop, 1)))
+
+/* Is DT_GPIO_CTLR(node_id, prop) the same node as DT_NODELABEL(gpio<idx>)? */
```
Instead of having a dependency on names, why don't we add a `bank` property to the GPIO node, with 0 for `NRF_P0` and 1 for `NRF_P1` and just multiply that by 32 as an offset?  That's also forward-looking to a potential `NRF_P2`.

mbolivar-nordic replied:
> Instead of having a dependency on names, why don't we add a `bank` property to the GPIO node, with 0 for `NRF_P0` and 1 for `NRF_P1` and just multiply that by 32 as an offset? That's also forward-looking to a potential `NRF_P2`.

Sure, I could do that if that's the consensus. This was just the easiest thing I could think of that didn't require going into all the .dtsi files.

Another alternative would be to match against `DT_REG_ADDR(node_id)` like `CHECK_DT_REG(gpio0, NRF_P0);` does in validate_base_addresses.c.

I'm not worried about NRF_P2 because that would require a new SoC port, at which point updating this file would be easy enough to do.

pabigot reviewed (commented):
Looks reasonable to me, though I'd prefer to have the GPIO controller ordinal provided as a property in the `nordic,nrf-gpio` node rather than inferred from node names.

For this:
> what "flags" value should we use for I2C sda-gpios and scl-gpios? The driver ignores them, so I guess maybe documenting that and saying "just put 0" is right, even if it's a bit inconsistent with Linux's use of those properties for bit-banged I2C.

Zero is correct in almost all cases.  `GPIO_PULL_UP` might be appropriate in some cases, though; I have used that when there was no external pull-up and it seemed to fix things.  If that's confirmed as a valid configuration then we might fix the drivers so that a non-zero flags corresponds to first initializing the pin to the specified state through the GPIO controller.

mbolivar-nordic commented:
@anangl @galak do you have an opinion on this part?

> what "flags" value should we use for I2C sda-gpios and scl-gpios? The driver ignores them, so I guess maybe documenting that and saying "just put 0" is right

If we agree on 0, I can keep going on this.

anangl commented:
> @anangl @galak do you have an opinion on this part?
> 
> > what "flags" value should we use for I2C sda-gpios and scl-gpios? The driver ignores them, so I guess maybe documenting that and saying "just put 0" is right
> 
> If we agree on 0, I can keep going on this.

I think this field could be used for specifying if the driver should enable pull-ups for those pins or not (if external pull-ups are present). If I understood correctly, @pabigot mentioned such functionality in the Dev Meeting. The nrfx drivers for TWI and TWIM do not currently provide such option (to specify if pull-ups are to be enabled), but this could be added quite easily, without breaking those driver APIs. Before this option is added, it could be e.g. checked in the shims with a build assertion if `GPIO_PULL_UP` flags were specified (as currently those pull-ups are always enabled by the nrfx drivers).

pabigot commented:
> If we agree on 0, I can keep going on this.

So from the generic I2C perspective my position is:

* It's reasonable to allow `0` rather than suggest people should use `GPIO_ACTIVE_whatever` (which is not relevant in this use; the driver/peripheral will always control direction configuration, even when the pin is configured as a GPIO).  IMO it's wrong to suggest `GPIO_PULL_UP` (because in almost all cases the pull-up will be external and the GPIO peripheral pulls would not be configured).  `0` should also be what gets synthesized if a `-pin` property is used.
* But I think it's reasonable to *allow* `GPIO_PULL_UP` if the intent is to rely on SOC internal pull-ups to eliminate the need for external ones.  Electrically I think this is not always reliable good idea, but in some cases it can work.

IMO it's not the place of this devicetree change to specify how the Nordic driver will interpret flags that aren't zero: whether GPIO-configured pull ups even affect I2C signals might be SOC-specific.  If/when we decided to do examine those flags we could document the behavior, but that should be done in the context of examining all I2C drivers to see if Nordic is the only one where this is even relevant.

If we need to document something I'd say "The flags cell may be 0.  The behavior if the flags cell that is not 0 is not yet defined."

mbolivar-nordic commented:
> If we need to document something I'd say "The flags cell may be 0. The behavior if the flags cell that is not 0 is not yet defined."

I'm fine with making nonzero values undefined behavior. @galak / @anangl -- agreed?

galak reviewed (commented):
Feels like some of the commits in here can be pulled out and merged.  Trying to look through and understand the need for the new DT APIs (DT_SAME_NODE) and (DT_NODE_PATH).

Need to look at the binding stuff and use of `FOO-gpios` v `pinctrl-N`.

mbolivar-nordic commented:
> Feels like some of the commits in here can be pulled out and merged.

OK, I've taken out the parts that aren't mergeable yet that demo this in the I2C drivers and just pushed the infrastructure.

Removed RFC and added 2.5 milestone. We do have a use case for this for Zephyr 2.5 (#30262) that will rebase on top once this is merged.

Please give it a real review if you can.

pabigot commented on dts/bindings/spi/spi-controller.yaml:
```diff
@@ -23,3 +24,33 @@ properties:
+        Example:
+
+          spi@... {
+                  cs-gpios = <&gpio0 pin0 flags0>,
```
If you make this actual values rather than symbols, e.g. `<&gpio0 23 GPIO_ACTIVE_LOW>` that may be more helpful.  Otherwise or in addition I think it's worth a comment saying that the flags value should almost certainly be `GPIO_ACTIVE_LOW` unless intervening hardware inverts the signal from what the peripheral expects.

Personal opinion, no change required.

pabigot reviewed (approved):
I'd still rather see a `bank` property in the devicetree node to provide the multiplier of 32 added to calculate the PSEL value, but that's just personal preference.

mbolivar-nordic commented:
> I'd still rather see a `bank` property in the devicetree node to
> provide the multiplier of 32 added to calculate the PSEL value, but
> that's just personal preference.

OK, I'll leave this as-is to save us some extra properties in each SoC
.dtsi for now, unless someone else speaks up and I become outnumbered on
this.

> If you make this actual values rather than symbols, e.g. `<&gpio0 23
> GPIO_ACTIVE_LOW>` that may be more helpful. Otherwise or in addition I
> think it's worth a comment saying that the flags value should almost
> certainly be `GPIO_ACTIVE_LOW` unless intervening hardware inverts the
> signal from what the peripheral expects.

Good ideas; both done.

mbolivar-nordic commented on soc/arm/nordic_nrf/soc_nrf_common.h:
```diff
@@ -0,0 +1,167 @@
+#define NRF_DT_GPIO_CTLR_IS_SOC(node_id, prop)				\
+	(NRF_DT_GPIO_CTLR_IS_GPIOX(node_id, prop, 0) ||			\
+	 UTIL_AND(DT_NODE_EXISTS(DT_NODELABEL(gpio1)),			\
+		  NRF_DT_GPIO_CTLR_IS_GPIOX(node_id, prop, 1)))
```
> Can't we just check here the compatible property of the controller node? I mean:

Of course. That's much better, thanks.

mbolivar-nordic commented:
@jfischer-no @anangl thanks for review. Comments addressed. The main changes are:

- dropped DT_SAME_NODE
- changed NRF_DT_CHECK_GPIO_CTLR_IS_SOC to use the compatible and removed obsolete private macros that used DT_SAME_NODE
- made NRF_PSEL_OFFSET_HELPER a requirement for soc.h that is used by soc_nrf_common.h; hopefully this helps future-proof the implementation
- each NRF_PSEL_OFFSET_HELPER now uses DT_REG_ADDR to distinguish nodes representing P0 from nodes representing P1
- re-worked spi-controller.yaml comments

mbolivar-nordic commented:
@galak :

> Feels like some of the commits in here can be pulled out and merged. Trying to look through and understand the need for the new DT APIs (DT_SAME_NODE) and (DT_NODE_PATH).

I dropped DT_SAME_NODE. The DT_NODE_PATH macro is just for making the error messages easier to read by including the 'real' paths to bad nodes.

> Need to look at the binding stuff and use of `FOO-gpios` v `pinctrl-N`.

I'm using `FOO-gpios` in anticipation of supporting upstream bindings like `sda-gpios` and `scl-gpios`: https://github.com/torvalds/linux/blob/c912fd05fab97934e4cf579654d0dc4835b4758c/Documentation/devicetree/bindings/i2c/i2c.txt#L79

ppryga commented:
@galak @jfischer-no gentlemen have you got something against merge of this PR? 
I'm waiting with my PR: https://github.com/zephyrproject-rtos/zephyr/pull/30262 util this one is merged.

mbolivar-nordic committed:
dts: nordic,nrf-gpio: add port property

This property is required information for extracting PSEL bits relevant
to peripheral pinmux from this node.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
Rebased to get through some unrelated compliance issues and cleaned up the commit logs.
